### PR TITLE
[779] Add new tool to create a full subsetting from Part Usage

### PR DESCRIPTION
**Pull request by jerome-obeo**
Bug: https://github.com/eclipse-syson/syson/issues/779

# PLEASE READ ALL ITEMS AND CHECK ONLY RELEVANT CHECKBOXES BELOW

## Project management

- [x] Has the pull request been added to the relevant milestone?
- [x] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [x] Have the relevant issues been added to the pull request?
- [x] Have the relevant labels been added to the issues? (`area:`, `type:`)
- [x] Have the relevant issues been added to the same project milestone as the pull request?

## Changelog and release notes

- [x] Has the `CHANGELOG.adoc` + `doc/content/modules/user-manual/pages/release-notes/YYYY.MM.0.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc` + `doc/content/modules/user-manual/pages/release-notes/YYYY.MM.0.adoc`?
- [ ] In case of a change with a visual impact, are there any screenshots in the `doc/content/modules/user-manual/pages/release-notes/YYYY.MM.0.adoc`?
- [ ] In case of a key change, has the change been added to `Key highlights` section in `doc/content/modules/user-manual/pages/release-notes/YYYY.MM.0.adoc`?
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc` + `doc/content/modules/user-manual/pages/release-notes/YYYY.MM.0.adoc`?

## Documentation

- [ ] Have you included an update of the [documentation](https://doc.mbse-syson.org) in your pull request? Please ask yourself if an update (installation manual, user manual, developer manual...) is needed and add one accordingly.

## Tests

- [x] Is the code properly tested? Any pull request (fix, enhancement or new feature) should come with a test (or several). It could be unit tests, integration tests or cypress tests depending on the context. Only doc and releng pull request do not need for tests.

### CODE REVIEW

File: CHANGELOG.adoc [hunk 1]
```diff
@@ -53,6 +53,7 @@ Rename the class `SemanticCheckerFactory` to `SemanticRunnableFactory` to reflec
 - https://github.com/eclipse-syson/syson/issues/771[#771] [diagrams] Allow the drop of elements on empty diagram nodes.
 - https://github.com/eclipse-syson/syson/issues/775[#775] [syson] Exclude `src/test/resources` from checkstyle.
 This reduces the time required to build SysON, especially when using Spring Tool Suite.
+- https://github.com/eclipse-syson/syson/issues/779[#779] [diagrams] Add a tool in Part usage to create a subsetting edge and the Part usage that is subset.
```
Comment (AxelRICHARD) [inline CHANGELOG.adoc:56]: "Part usage" -> "`PartUsage`" (note the "`")

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewCreateService.java [hunk 2]
```diff
@@ -950,4 +956,31 @@ private boolean isPart(Element element) {
...
+
+    /**
+     * Creates a new sibling part usage as well as a subsetting edge between those parts.
+     *
+     * @param self
+     *         the {@link PartUsage} to subset by a new part usage
+     * @return the new part usage or self if something went wrong.
+     */
+    public Object createPartUsageAndSubsetting(Element self) {
+        if (self instanceof PartUsage partUsage) {
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewCreateService.java:968]: It seems this service can only be applied on `PartUsage`. So could you check that the following signature: `public PartUsage createPartUsageAndSubsetting(PartUsage self)` is ok? Among your tests, please verify that the backend console does not display an AQL error with this new signature.

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageSubsettingNodeToolProvider.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    public PartUsageSubsettingNodeToolProvider(IDescriptionNameGenerator nameGenerator) {
+        this.nameGenerator = Objects.requireNonNull(nameGenerator);
+    }
+
+    @Override
+    public NodeTool create(IViewDiagramElementFinder cache) {
+        var builder = this.diagramBuilderHelper.newNodeTool();
+
+        var creationSubsettingServiceCall = this.viewBuilderHelper.newChangeContext()
+                .expression(AQLConstants.AQL_SELF + ".createPartUsageAndSubsetting()")
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageSubsettingNodeToolProvider.java:53]: Please prefer `AQLUtils.getSelfService...`

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageSubsettingNodeToolProvider.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                .build();
+
+        NodeDescription nodeDescription = cache.getNodeDescription(nameGenerator.getNodeName(SysmlPackage.eINSTANCE.getPartUsage())).orElse(null);
+
+        var createView = this.diagramBuilderHelper.newCreateView()
+                .containmentKind(NodeContainmentKind.CHILD_NODE)
+                .elementDescription(nodeDescription)
+                .parentViewExpression(AQLUtils.getSelfServiceCallExpression("getParentNode", List.of("selectedNode", "diagramContext")))
+                .semanticElementExpression("aql:self")
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageSubsettingNodeToolProvider.java:62]: Please prefer AQLConstants.AQL_SELF

File: CHANGELOG.adoc (hunk 1 above)
Comment (AxelRICHARD) [reply CHANGELOG.adoc:56]: and the Part usage -> and a `PartUsage`

Comment (AxelRICHARD) [reply CHANGELOG.adoc:56]: subsetting -> `Subsetting`

Comment (AxelRICHARD) [reply CHANGELOG.adoc:56]: tool in -> tool on

Comment (AxelRICHARD) [reply CHANGELOG.adoc:56]: Could you also add the name of this new tool in this message please?

Comment (jerome-obeo) [reply CHANGELOG.adoc:56]: Done

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageSubsettingNodeToolProvider.java (hunk 3 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageSubsettingNodeToolProvider.java:53]: Done

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageSubsettingNodeToolProvider.java (hunk 4 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageSubsettingNodeToolProvider.java:62]: Done

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewCreateService.java (hunk 2 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewCreateService.java:968]: Done.
There are no messages when the signature is changed.

**Review by AxelRICHARD — APPROVED** [review]

**Pull request merged**